### PR TITLE
[AMORO-2710] Support bulk operations for arctic file io

### LINKED ISSUE #2710: [Improvement]: The time overhead for file deletion during maintenance is significant
### Search before asking

- [X] I have searched in the [issues](https://github.com/NetEase/amoro/issues?q=is%3Aissue) and found no similar issues.

### What would you like to be improved?

Reduce the time overhead of deletion during maintenance (Mixed and Iceberg format)

### How should we improve?

1. Support `SupportsBulkOperations` if needed
2. Batch delete files, it is very useful for S3 delete operations. refer to https://github.com/apache/iceberg/pull/4052

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Subtasks

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by XBaith**
## Why are the changes needed?

Close #2710

## Brief change log

- To improve the efficiency of file deletion and avoid rate limiting due to a large number of requests, especially for S3

## How was this patch tested?

- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

- Does this pull request introduce a new feature? (no)
- If yes, how is the feature documented? (not documented)

### CODE REVIEW

File: ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java [hunk 1]
```diff
@@ -297,21 +300,32 @@ protected void cleanContentFiles(long lastTime) {
...
     // to prevent repeated acquisition
     Set<String> validFiles = orphanFileCleanNeedToExcludeFiles();
-    LOG.info("{} start clean content files of change store", table.name());
+    LOG.info("{} start cleaning orphan files in content", table.name());
     int deleteFilesCnt = clearInternalTableContentsFiles(lastTime, validFiles);
-    LOG.info("{} total delete {} files from change store", table.name(), deleteFilesCnt);
+    runWithCondition(
+        deleteFilesCnt > 0,
+        () -> LOG.info("{} total delete {} orphan files in content", table.name(), deleteFilesCnt),
+        () -> LOG.info("{} doesn't have orphan files in content", table.name()));
```
[the remark this answers is not in the document]
Comment (XBaith) [reply ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java:308]: OKay, my goal at the beginning was to reduce unnecessary logging, so what do you think about not printing the operation that deletes 0 files

[the remark this answers is not in the document]
Comment (baiyangtx) [reply ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java:308]: I think if there is no files need to be deleted, it's fine to not logging, but if it has some files need to be deleted and the actual deleted file is 0, in this case the logging is needed.

so could you add the `files need to be deleted` and `actual deleted files` to the logging?

File: ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java [hunk 2]
```diff
@@ -495,28 +509,31 @@ private static int deleteInvalidFilesInFs(
...
+        Set<String> filesToDelete = new HashSet<>();
         if (!excludes.contains(uriPath)
             && !excludes.contains(parentUriPath)
             && p.createdAtMillis() < lastTime) {
-          fio.deleteFile(p.location());
-          deleteCount += 1;
+          filesToDelete.add(p.location());
         }
+        int deletedCnt =
+            TableFileUtil.deleteFiles(table.name(), arcticFileIO(), filesToDelete, false);
```
[the remark this answers is not in the document]
Comment (XBaith) [reply ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java]: Thanks, I update code base on your comment, PTAL @baiyangtx

File: ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java [hunk 3]
```diff
@@ -352,10 +368,18 @@ private int clearInternalTableContentsFiles(long lastTime, Set<String> exclude)
...
       if (io.supportFileSystemOperations()) {
         SupportsFileSystemOperations fio = io.asFileSystemIO();
-        return deleteInvalidFilesInFs(fio, dataLocation, lastTime, exclude);
+        Set<PathInfo> directories = new HashSet<>();
+        Set<String> filesToDelete =
+            deleteInvalidFilesInFs(fio, dataLocation, lastTime, exclude, directories);
+        int deleted = TableFileUtil.deleteFiles(io, filesToDelete);
+        /* delete empty directories */
+        deleteEmptyDirectories(fio, directories, lastTime, exclude);
+        return deleted;
```
Comment (baiyangtx) [inline ams/server/src/main/java/com/netease/arctic/server/optimizing/maintainer/IcebergTableMaintainer.java:377]: I means that we should print log like this.

```
runWithCondition( 
filesToDelete.size() > 0 , 
() -> LOG.info("{}: There are {} files need to be deleted, and delete {} files successful.", table.name(), filesToDelete.size(), deleted)); 

```

File: core/src/main/java/com/netease/arctic/utils/TableFileUtil.java [hunk 4]
```diff
@@ -80,6 +84,81 @@ public static void deleteEmptyDirectory(
...
+  /**
+   * Helper to delete files. Bulk deletion is used if possible.
+   *
+   * @param io arctic file io
+   * @param files files to delete
+   * @param concurrent controls concurrent deletion. Only applicable for non-bulk FileIO
+   * @return deleted file count
+   */
+  public static int deleteFiles(
+      ArcticFileIO io, Set<String> files, boolean concurrent, ExecutorService svc) {
```
[the remark this answers is not in the document]
Comment (XBaith) [reply core/src/main/java/com/netease/arctic/utils/TableFileUtil.java]: This way the code will be cleaner, thanks

**Review by baiyangtx — APPROVED** [review]
LGTM, thanks for your contributions.

**Pull request merged**